File: tools/testing/connection/undici/figma.mjs
/**
 * Figma API - undici Connection Test
 *
 * Authentication: X-Figma-Token header
 * Base URL: https://api.figma.com/v1
 * Health Endpoint: GET /v1/me
 *
 * TLS/SSL Options:
 *   NODE_TLS_REJECT_UNAUTHORIZED=0  - Ignore all certificate errors
 *   REQUEST_CA_BUNDLE=/path/to/ca   - Custom CA bundle file
 *   SSL_CERT_FILE=/path/to/cert     - Custom SSL certificate file
 *   NODE_EXTRA_CA_CERTS=/path/to/ca - Additional CA certificates
 */

import { request, ProxyAgent, Agent } from 'undici';
import { readFileSync } from 'node:fs';

// ============================================================================
// Configuration - Override these values
// ============================================================================

const CONFIG = {
  // Required
  FIGMA_TOKEN: process.env.FIGMA_TOKEN || 'your_figma_token_here',

  // Base URL
  BASE_URL: 'https://api.figma.com',

  // Optional: Proxy Configuration
  HTTPS_PROXY: process.env.HTTPS_PROXY || '', // e.g., 'http://proxy.example.com:8080'

  // Optional: TLS Configuration
  // Set NODE_TLS_REJECT_UNAUTHORIZED=0 to ignore certificate errors
  REJECT_UNAUTHORIZED: process.env.NODE_TLS_REJECT_UNAUTHORIZED !== '0',

  // Optional: Custom CA certificates
  // REQUEST_CA_BUNDLE or SSL_CERT_FILE - path to custom CA bundle
  // NODE_EXTRA_CA_CERTS - handled automatically by Node.js
  CA_BUNDLE: process.env.REQUEST_CA_BUNDLE || process.env.SSL_CERT_FILE || '',
};

// ============================================================================
// TLS Configuration Helper
// ============================================================================

function getTlsConnectOptions() {
  const options = {
    rejectUnauthorized: CONFIG.REJECT_UNAUTHORIZED,
  };

  // Load custom CA bundle if specified
  if (CONFIG.CA_BUNDLE) {
    try {
      options.ca = readFileSync(CONFIG.CA_BUNDLE);
      console.log(`Using custom CA bundle: ${CONFIG.CA_BUNDLE}`);
    } catch (err) {
      console.warn(`Warning: Could not read CA bundle: ${err.message}`);
    }
  }

  return options;
}

// ============================================================================
// Create Dispatcher (with or without proxy)
// ============================================================================

function createDispatcher() {
  const tlsOptions = getTlsConnectOptions();

  if (CONFIG.HTTPS_PROXY) {
    console.log(`Using proxy: ${CONFIG.HTTPS_PROXY}`);
    return new ProxyAgent({
      uri: CONFIG.HTTPS_PROXY,
      connect: tlsOptions,
    });
  }
  return new Agent({
    connect: tlsOptions,
  });
}

// ============================================================================
// Health Check
// ============================================================================

async function healthCheck() {
  console.log('\n=== Figma Health Check ===\n');

  const dispatcher = createDispatcher();

  try {
    const response = await request(`${CONFIG.BASE_URL}/v1/me`, {
      method: 'GET',
      headers: {
        'X-Figma-Token': CONFIG.FIGMA_TOKEN,
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log('Response:', JSON.stringify(data, null, 2));

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

// ============================================================================
// Sample API Calls
// ============================================================================

async function getFile(fileKey) {
  console.log(`\n=== Get File: ${fileKey} ===\n`);

  const dispatcher = createDispatcher();

  try {
    const response = await request(`${CONFIG.BASE_URL}/v1/files/${fileKey}`, {
      method: 'GET',
      headers: {
        'X-Figma-Token': CONFIG.FIGMA_TOKEN,
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log('File Name:', data.name);
    console.log('Last Modified:', data.lastModified);

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

async function getFileNodes(fileKey, nodeIds) {
  console.log(`\n=== Get File Nodes: ${fileKey} ===\n`);

  const dispatcher = createDispatcher();

  try {
    const url = new URL(`${CONFIG.BASE_URL}/v1/files/${fileKey}/nodes`);
    url.searchParams.set('ids', nodeIds.join(','));

    const response = await request(url.toString(), {
      method: 'GET',
      headers: {
        'X-Figma-Token': CONFIG.FIGMA_TOKEN,
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log('Response:', JSON.stringify(data, null, 2));

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

async function getFileImages(fileKey, nodeIds, format = 'png') {
  console.log(`\n=== Get File Images: ${fileKey} ===\n`);

  const dispatcher = createDispatcher();

  try {
    const url = new URL(`${CONFIG.BASE_URL}/v1/images/${fileKey}`);
    url.searchParams.set('ids', nodeIds.join(','));
    url.searchParams.set('format', format);

    const response = await request(url.toString(), {
      method: 'GET',
      headers: {
        'X-Figma-Token': CONFIG.FIGMA_TOKEN,
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log('Images:', data.images);

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

async function listTeamProjects(teamId) {
  console.log(`\n=== List Team Projects: ${teamId} ===\n`);

  const dispatcher = createDispatcher();

  try {
    const response = await request(`${CONFIG.BASE_URL}/v1/teams/${teamId}/projects`, {
      method: 'GET',
      headers: {
        'X-Figma-Token': CONFIG.FIGMA_TOKEN,
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log(`Found ${data.projects?.length || 0} projects`);
    data.projects?.forEach((project) => {
      console.log(`  - ${project.name} (${project.id})`);
    });

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

// ============================================================================
// Run Tests
// ============================================================================

async function main() {
  console.log('Figma API Connection Test');
  console.log('=========================');
  console.log(`Base URL: ${CONFIG.BASE_URL}`);
  console.log(`Proxy: ${CONFIG.HTTPS_PROXY || 'None'}`);
  console.log(`Token: ${CONFIG.FIGMA_TOKEN.slice(0, 10)}...`);
  console.log(`TLS Verify: ${CONFIG.REJECT_UNAUTHORIZED}`);
  console.log(`CA Bundle: ${CONFIG.CA_BUNDLE || 'System default'}`);

  await healthCheck();
  // await getFile('file_key_here');
  // await getFileNodes('file_key_here', ['node_id_1', 'node_id_2']);
  // await getFileImages('file_key_here', ['node_id_1'], 'png');
  // await listTeamProjects('team_id_here');
}

main().catch(console.error);

File: tools/testing/connection/httpx/figma.py
#!/usr/bin/env python3
"""
Figma API - httpx Connection Test

Authentication: X-Figma-Token header
Base URL: https://api.figma.com/v1
Health Endpoint: GET /v1/me

TLS/SSL Options:
  SSL_CERT_VERIFY=0              - Ignore all certificate errors
  REQUEST_CA_BUNDLE=/path/to/ca  - Custom CA bundle file
  SSL_CERT_FILE=/path/to/cert    - Custom SSL certificate file
  REQUESTS_CA_BUNDLE=/path/to/ca - Alternative CA bundle (requests compat)
"""

import asyncio
import json
import os
from typing import Any, Union

import httpx

# ============================================================================
# Configuration - Override these values
# ============================================================================

CONFIG = {
    # Required
    "FIGMA_TOKEN": os.getenv("FIGMA_TOKEN", "your_figma_token_here"),

    # Base URL
    "BASE_URL": "https://api.figma.com",

    # Optional: Proxy Configuration
    "HTTPS_PROXY": os.getenv("HTTPS_PROXY", ""),  # e.g., "http://proxy.example.com:8080"

    # Optional: TLS Configuration
    # Set to False to ignore certificate errors (default: False for testing)
    # SSL_CERT_VERIFY=0, REQUEST_CA_BUNDLE=null, SSL_CERT_FILE=null, REQUESTS_CA_BUNDLE=null
    "VERIFY_SSL": False,

    # Optional: Custom CA certificates (disabled by default for testing)
    "CA_BUNDLE": None,
}


# ============================================================================
# TLS Configuration Helper
# ============================================================================

def get_ssl_verify() -> Union[bool, str]:
    """Get SSL verification setting."""
    if CONFIG["CA_BUNDLE"]:
        print(f"Using custom CA bundle: {CONFIG['CA_BUNDLE']}")
        return CONFIG["CA_BUNDLE"]
    return CONFIG["VERIFY_SSL"]


# ============================================================================
# Create HTTP Client
# ============================================================================

def create_client() -> httpx.AsyncClient:
    """Create httpx async client with optional proxy."""
    proxies = None
    if CONFIG["HTTPS_PROXY"]:
        print(f"Using proxy: {CONFIG['HTTPS_PROXY']}")
        proxies = {
            "http://": CONFIG["HTTPS_PROXY"],
            "https://": CONFIG["HTTPS_PROXY"],
        }

    return httpx.AsyncClient(
        proxies=proxies,
        verify=get_ssl_verify(),
        timeout=httpx.Timeout(30.0),
    )


# ============================================================================
# Health Check
# ============================================================================

async def health_check() -> dict[str, Any]:
    """Perform health check against Figma API."""
    print("\n=== Figma Health Check ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{CONFIG['BASE_URL']}/v1/me",
                headers={
                    "X-Figma-Token": CONFIG["FIGMA_TOKEN"],
                },
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            print(f"Response: {json.dumps(data, indent=2)}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


# ============================================================================
# Sample API Calls
# ============================================================================

async def get_file(file_key: str) -> dict[str, Any]:
    """Get file details."""
    print(f"\n=== Get File: {file_key} ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{CONFIG['BASE_URL']}/v1/files/{file_key}",
                headers={
                    "X-Figma-Token": CONFIG["FIGMA_TOKEN"],
                },
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            print(f"File Name: {data.get('name')}")
            print(f"Last Modified: {data.get('lastModified')}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


async def get_file_nodes(file_key: str, node_ids: list[str]) -> dict[str, Any]:
    """Get specific nodes from a file."""
    print(f"\n=== Get File Nodes: {file_key} ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{CONFIG['BASE_URL']}/v1/files/{file_key}/nodes",
                params={"ids": ",".join(node_ids)},
                headers={
                    "X-Figma-Token": CONFIG["FIGMA_TOKEN"],
                },
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            print(f"Response: {json.dumps(data, indent=2)}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


async def get_file_images(file_key: str, node_ids: list[str], format: str = "png") -> dict[str, Any]:
    """Get images for specific nodes."""
    print(f"\n=== Get File Images: {file_key} ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{CONFIG['BASE_URL']}/v1/images/{file_key}",
                params={"ids": ",".join(node_ids), "format": format},
                headers={
                    "X-Figma-Token": CONFIG["FIGMA_TOKEN"],
                },
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            print(f"Images: {data.get('images')}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


async def list_team_projects(team_id: str) -> dict[str, Any]:
    """List projects in a team."""
    print(f"\n=== List Team Projects: {team_id} ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{CONFIG['BASE_URL']}/v1/teams/{team_id}/projects",
                headers={
                    "X-Figma-Token": CONFIG["FIGMA_TOKEN"],
                },
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            projects = data.get("projects", [])
            print(f"Found {len(projects)} projects")
            for project in projects:
                print(f"  - {project['name']} ({project['id']})")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


# ============================================================================
# Run Tests
# ============================================================================

async def main():
    """Run connection tests."""
    print("Figma API Connection Test")
    print("=========================")
    print(f"Base URL: {CONFIG['BASE_URL']}")
    print(f"Proxy: {CONFIG['HTTPS_PROXY'] or 'None'}")
    print(f"Token: {CONFIG['FIGMA_TOKEN'][:10]}...")
    print(f"SSL Verify: {CONFIG['VERIFY_SSL']}")
    print(f"CA Bundle: {CONFIG['CA_BUNDLE'] or 'System default'}")

    await health_check()
    # await get_file("file_key_here")
    # await get_file_nodes("file_key_here", ["node_id_1", "node_id_2"])
    # await get_file_images("file_key_here", ["node_id_1"], "png")
    # await list_team_projects("team_id_here")


if __name__ == "__main__":
    asyncio.run(main())
